Assets/BossRoom/Scripts/Client/Game/State/ClientMainMenuState.cs

```diff
@@ -11,6 +11,50 @@ namespace BossRoom.Client
     /// </summary>
     public class ClientMainMenuState : GameStateBehaviour
     {
+        /// <summary>
+        /// Used in concert with SetTransitionReason to let code that triggers transition to the MainMenu scene set context
+        /// for why the transition occurred. This is then used to display an appropriate message to the user in the UI.
+        /// </summary>
+        public enum TransitionReason
+        {
+            Undefined,      //no reason has been set. 
+            UserRequested,  //user explicitly requested a disconnect. 
+            Disconnect,     //client unexpectedly lost connection with host. 
+        }
+
+        private static TransitionReason s_TransitionReason;
+
+        /// <summary>
+        /// Set this to have the MainMenu display a MessageBox to the user on re-entering the MainMenu. Useful for displaying error states. 
+        /// </summary>
+        public static void SetTransitionReason( TransitionReason reason )
+        {
+            s_TransitionReason = reason;
+        }
+
+        /// <summary>
+        /// Has a TransitionReason already be set? (The TransitionReason provides context for why someone transition back to the MainMenu, and is a one-use item
+        /// that is unset as soon as it is read).
+        /// </summary>
+        public static bool HasTransitionReason => s_TransitionReason != TransitionReason.Undefined;
+
+        /// <summary>
+        /// MainMenuUI should invoke this on start, and display a message to the user if not Undefined
+        /// </summary>
+        /// <returns>TransitionReason set by whoever triggered the scene transition to main menu, or undefined if none. </returns>
+        public TransitionReason ReadAndUnsetTransitionReason()
+        {
+            var reason = s_TransitionReason;
+            s_TransitionReason = TransitionReason.Undefined;
+            return reason;
+        }
```

Comment:
disconnect reason should be put in game net portal, non static

Reply:
Personally, I would encapsulate this in it's own struct/class

Reply:
We have a preference to keep it with GameNetPortal because we may some day have multiple active connections, with multiple GameNetPortals, each tracking the state of a connection.

Reply:
I've completed this update. There's now a `DisconnectReason` class that hangs off ClientGameNetPortal. The state lasts until the client starts another connection (or becomes a host).

Reply:
Looks great. Thanks.